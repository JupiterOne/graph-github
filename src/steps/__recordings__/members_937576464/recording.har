{
  "log": {
    "_recordingName": "members",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "06d6593fc4a07c98683e9f0972b1c5e0",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/vnd.github.machine-man-preview+json"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "jupiter-integration-github octokit-rest.js/18.12.0 octokit-core.js/3.5.1 Node.js/16.14.2 (darwin; x64)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.github.com"
            }
          ],
          "headersSize": 813,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "queryString": [],
          "url": "https://api.github.com/app/installations/23522163/access_tokens"
        },
        "response": {
          "bodySize": 102,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 102,
            "text": "{\"message\":\"A JSON web token could not be decoded\",\"documentation_url\":\"https://docs.github.com/rest\",\"token\":\"[REDACTED]\",\"expires_at\":\"2050-12-31T18:09:20Z\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "GitHub.com"
            },
            {
              "name": "date",
              "value": "Fri, 02 Sep 2022 15:59:41 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "content-length",
              "value": "102"
            },
            {
              "name": "x-github-media-type",
              "value": "github.v3; param=machine-man-preview; format=json"
            },
            {
              "name": "access-control-expose-headers",
              "value": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubdomains; preload"
            },
            {
              "name": "x-frame-options",
              "value": "deny"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "referrer-policy",
              "value": "origin-when-cross-origin, strict-origin-when-cross-origin"
            },
            {
              "name": "content-security-policy",
              "value": "default-src 'none'"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, Accept, X-Requested-With"
            },
            {
              "name": "x-github-request-id",
              "value": "F4B9:689B:57D06:D196E:6312286D"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 939,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 401,
          "statusText": "Unauthorized"
        },
        "startedDateTime": "2022-09-02T15:59:40.909Z",
        "time": 160,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 160
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
