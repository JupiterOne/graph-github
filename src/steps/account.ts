import {
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../config';
import { createAPIClient } from '../client';
import { toAccountEntity } from '../sync/converters';
import {
  GITHUB_ACCOUNT_ENTITY_TYPE,
  GITHUB_ACCOUNT_ENTITY_CLASS,
} from '../constants';

export const DATA_ACCOUNT_ENTITY = 'DATA_ACCOUNT_ENTITY';

export async function fetchAccountDetails({
  jobState,
  instance,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const config = instance.config;
  const apiClient = createAPIClient(config, logger);
  const accountEntity = await jobState.addEntity(
    toAccountEntity(await apiClient.getAccountDetails()),
  );

  await jobState.setData(DATA_ACCOUNT_ENTITY, accountEntity);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'Account',
        _type: GITHUB_ACCOUNT_ENTITY_TYPE,
        _class: GITHUB_ACCOUNT_ENTITY_CLASS,
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
