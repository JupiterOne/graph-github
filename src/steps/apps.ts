import {
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  IntegrationMissingKeyError,
  createDirectRelationship,
} from '@jupiterone/integration-sdk-core';

import { getOrCreateApiClient } from '../client';
import { IntegrationConfig } from '../config';
import { DATA_ACCOUNT_ENTITY } from './account';
import { toAppEntity } from '../sync/converters';
import { AccountEntity, AppEntity } from '../types';
import {
  GithubEntities,
  GITHUB_ACCOUNT_APP_RELATIONSHIP_TYPE,
  GITHUB_APP_BY_APP_ID,
  Steps,
} from '../constants';

export async function fetchApps({
  instance,
  logger,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const config = instance.config;
  const apiClient = getOrCreateApiClient(config, logger);

  const accountEntity = await jobState.getData<AccountEntity>(
    DATA_ACCOUNT_ENTITY,
  );

  if (!accountEntity) {
    throw new IntegrationMissingKeyError(
      `Expected to find Account entity in jobState.`,
    );
  }

  const appIdMap = {};
  await apiClient.iterateApps(async (app) => {
    const appEntity = (await jobState.addEntity(toAppEntity(app))) as AppEntity;

    appIdMap[app.app_id] = appEntity;

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.INSTALLED,
        from: accountEntity,
        to: appEntity,
      }),
    );
  });

  await jobState.setData(GITHUB_APP_BY_APP_ID, appIdMap);
}

export const appSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: Steps.FETCH_APPS,
    name: 'Fetch Apps',
    entities: [
      {
        resourceName: 'Github App',
        _type: GithubEntities.GITHUB_APP._type,
        _class: GithubEntities.GITHUB_APP._class,
      },
    ],
    relationships: [
      {
        _type: GITHUB_ACCOUNT_APP_RELATIONSHIP_TYPE,
        _class: RelationshipClass.INSTALLED,
        sourceType: GithubEntities.GITHUB_ACCOUNT._type,
        targetType: GithubEntities.GITHUB_APP._type,
      },
    ],
    dependsOn: [Steps.FETCH_ACCOUNT],
    executionHandler: fetchApps,
  },
];
