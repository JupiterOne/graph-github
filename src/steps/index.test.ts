import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig, sanitizeConfig } from '../config';
import { fetchMembers } from './members';
import { fetchRepos } from './repos';
import { fetchTeams } from './teams';
import { fetchCollaborators } from './collaborators';
import { fetchPrs } from './pullRequests';
import { fetchAccountDetails } from './account';
import { fetchApps } from './apps';
import { GITHUB_REPO_SECRET_ORG_SECRET_RELATIONSHIP_TYPE } from '../constants';
import { integrationConfig } from '../../test/config';
import { setupGithubRecording } from '../../test/recording';
import { fetchOrgSecrets } from './orgSecrets';
import { fetchRepoSecrets } from './repoSecrets';
import { fetchEnvironments } from './environments';
import { fetchIssues } from './issues';
import { fetchEnvSecrets } from './envSecrets';
import { fetchTeamRepos } from './teamRepos';
import { fetchTeamMembers } from './teamMembers';

jest.setTimeout(75000);

let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

test('should collect data', async () => {
  recording = setupGithubRecording({
    directory: __dirname,
    name: 'steps', //redaction of headers is in setupGithubRecording
  });

  sanitizeConfig(integrationConfig);
  integrationConfig.installationId = 17214088; //this is the id the recordings are under

  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  await fetchApps(context);
  await fetchMembers(context);
  await fetchRepos(context);
  await fetchTeams(context);
  await fetchTeamRepos(context);
  await fetchTeamMembers(context);
  await fetchCollaborators(context);
  await fetchPrs(context);
  await fetchIssues(context);
  await fetchEnvironments(context);
  await fetchOrgSecrets(context);
  await fetchRepoSecrets(context);
  await fetchEnvSecrets(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const userGroups = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('UserGroup'),
  );
  expect(userGroups.length).toBeGreaterThan(0);
  expect(userGroups).toMatchGraphObjectSchema({
    _class: ['UserGroup'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'github_team' },
        webLink: { type: 'string' },
        displayName: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['webLink', 'displayName'],
    },
  });

  const repos = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('CodeRepo'),
  );
  expect(repos.length).toBeGreaterThan(0);
  expect(repos).toMatchGraphObjectSchema({
    _class: ['CodeRepo'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'github_repo' },
        webLink: { type: 'string' },
        displayName: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['webLink', 'displayName'],
    },
  });

  const repoSecrets = context.jobState.collectedEntities.filter(
    (e) =>
      e._class.includes('Secret') && e._type.includes('github_repo_secret'),
  );
  expect(repoSecrets.length).toBeGreaterThan(0);
  expect(repoSecrets).toMatchGraphObjectSchema({
    _class: ['Secret'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'github_repo_secret' },
        webLink: { type: 'string' },
        displayName: { type: 'string' },
        name: { type: 'string' },
        createdOn: { type: 'number' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['webLink', 'displayName', 'name', 'createdOn'],
    },
  });

  const secretRepoOrgOverrideRelationships =
    context.jobState.collectedRelationships.filter(
      (r) => r._type === GITHUB_REPO_SECRET_ORG_SECRET_RELATIONSHIP_TYPE,
    );
  expect(secretRepoOrgOverrideRelationships.length).toBeGreaterThan(0);
});
