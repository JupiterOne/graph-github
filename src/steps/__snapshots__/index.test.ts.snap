// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
            "login": "Kei-Institute",
            "name": "Kei-Institute",
          },
        },
      ],
      "_type": "github_account",
      "accountId": "Kei-Institute",
      "accountType": "Organization",
      "displayName": "Kei-Institute",
      "login": "Kei-Institute",
      "name": "Kei-Institute",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjUxMzUyMw==",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjUxMzUyMw==",
            "isSiteAdmin": false,
            "login": "erichs",
            "name": "Erich Smith",
            "node": undefined,
            "organization": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
            "role": "MEMBER",
          },
        },
      ],
      "_type": "github_user",
      "displayName": "Erich Smith",
      "login": "erichs",
      "mfaEnabled": false,
      "name": "Erich Smith",
      "role": "MEMBER",
      "siteAdmin": false,
      "username": "erichs",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjI1NDg5NDgy",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjI1NDg5NDgy",
            "isSiteAdmin": false,
            "login": "mknoedel",
            "name": "Michael Knoedel",
            "node": undefined,
            "organization": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
            "role": "MEMBER",
          },
        },
      ],
      "_type": "github_user",
      "displayName": "Michael Knoedel",
      "login": "mknoedel",
      "mfaEnabled": false,
      "name": "Michael Knoedel",
      "role": "MEMBER",
      "siteAdmin": false,
      "username": "mknoedel",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "MDQ6VXNlcjYyNDkyMDk3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "hasTwoFactorEnabled": null,
            "id": "MDQ6VXNlcjYyNDkyMDk3",
            "isSiteAdmin": false,
            "login": "kevincasey1222",
            "name": "Kevin Casey",
            "node": undefined,
            "organization": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
            "role": "ADMIN",
          },
        },
      ],
      "_type": "github_user",
      "displayName": "Kevin Casey",
      "login": "kevincasey1222",
      "mfaEnabled": false,
      "name": "Kevin Casey",
      "role": "ADMIN",
      "siteAdmin": false,
      "username": "kevincasey1222",
    },
    Object {
      "_class": Array [
        "CodeRepo",
      ],
      "_key": "MDEwOlJlcG9zaXRvcnkzNzE0MTk1OTg=",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createdAt": "2021-05-27T15:23:24Z",
            "id": "MDEwOlJlcG9zaXRvcnkzNzE0MTk1OTg=",
            "isArchived": false,
            "isPrivate": false,
            "name": "Test-repo",
            "nameWithOwner": "Kei-Institute/Test-repo",
            "node": undefined,
            "organization": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
            "updatedAt": "2021-05-27T15:23:28Z",
            "url": "https://github.com/Kei-Institute/Test-repo",
          },
        },
      ],
      "_type": "github_repo",
      "archived": false,
      "createdOn": 1622129004000,
      "displayName": "Test-repo",
      "fullName": "kei-institute/test-repo",
      "name": "Test-repo",
      "owner": "kei-institute",
      "public": true,
      "updatedOn": 1622129008000,
      "webLink": "https://github.com/Kei-Institute/Test-repo",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "MDQ6VGVhbTQ4NTgxNjk=",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "MDQ6VGVhbTQ4NTgxNjk=",
            "members": Array [
              Object {
                "id": "MDQ6VXNlcjUxMzUyMw==",
                "login": "erichs",
                "node": undefined,
                "role": "MEMBER",
                "teams": "MDQ6VGVhbTQ4NTgxNjk=",
              },
              Object {
                "id": "MDQ6VXNlcjI1NDg5NDgy",
                "login": "mknoedel",
                "node": undefined,
                "role": "MEMBER",
                "teams": "MDQ6VGVhbTQ4NTgxNjk=",
              },
              Object {
                "id": "MDQ6VXNlcjYyNDkyMDk3",
                "login": "kevincasey1222",
                "node": undefined,
                "role": "MAINTAINER",
                "teams": "MDQ6VGVhbTQ4NTgxNjk=",
              },
            ],
            "name": "Betterteam",
            "node": undefined,
            "organization": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
            "repos": Array [],
            "slug": "betterteam",
            "url": "https://github.com/orgs/Kei-Institute/teams/betterteam",
          },
        },
      ],
      "_type": "github_team",
      "displayName": "Betterteam",
      "fullName": "Betterteam",
      "name": "betterteam",
      "webLink": "https://github.com/orgs/Kei-Institute/teams/betterteam",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "MDQ6VGVhbTQ4NTgxNzA=",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "MDQ6VGVhbTQ4NTgxNzA=",
            "members": Array [
              Object {
                "id": "MDQ6VXNlcjYyNDkyMDk3",
                "login": "kevincasey1222",
                "node": undefined,
                "role": "MAINTAINER",
                "teams": "MDQ6VGVhbTQ4NTgxNzA=",
              },
            ],
            "name": "Childteam",
            "node": undefined,
            "organization": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
            "repos": Array [],
            "slug": "childteam",
            "url": "https://github.com/orgs/Kei-Institute/teams/childteam",
          },
        },
      ],
      "_type": "github_team",
      "displayName": "Childteam",
      "fullName": "Childteam",
      "name": "childteam",
      "webLink": "https://github.com/orgs/Kei-Institute/teams/childteam",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "MDQ6VGVhbTQ4NTc0OTU=",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "MDQ6VGVhbTQ4NTc0OTU=",
            "members": Array [
              Object {
                "id": "MDQ6VXNlcjI1NDg5NDgy",
                "login": "mknoedel",
                "node": undefined,
                "role": "MEMBER",
                "teams": "MDQ6VGVhbTQ4NTc0OTU=",
              },
              Object {
                "id": "MDQ6VXNlcjYyNDkyMDk3",
                "login": "kevincasey1222",
                "node": undefined,
                "role": "MAINTAINER",
                "teams": "MDQ6VGVhbTQ4NTc0OTU=",
              },
            ],
            "name": "Test team",
            "node": undefined,
            "organization": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
            "repos": Array [],
            "slug": "test-team",
            "url": "https://github.com/orgs/Kei-Institute/teams/test-team",
          },
        },
      ],
      "_type": "github_team",
      "displayName": "Test team",
      "fullName": "Test team",
      "name": "test-team",
      "webLink": "https://github.com/orgs/Kei-Institute/teams/test-team",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_key": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz|has|MDQ6VXNlcjUxMzUyMw==",
      "_toEntityKey": "MDQ6VXNlcjUxMzUyMw==",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_key": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz|has|MDQ6VXNlcjI1NDg5NDgy",
      "_toEntityKey": "MDQ6VXNlcjI1NDg5NDgy",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_key": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz|has|MDQ6VXNlcjYyNDkyMDk3",
      "_toEntityKey": "MDQ6VXNlcjYyNDkyMDk3",
      "_type": "github_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjYyNDkyMDk3",
      "_key": "MDQ6VXNlcjYyNDkyMDk3|manages|MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_toEntityKey": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_type": "github_user_manages_account",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "OWNS",
      "_fromEntityKey": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_key": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz|owns|MDEwOlJlcG9zaXRvcnkzNzE0MTk1OTg=",
      "_toEntityKey": "MDEwOlJlcG9zaXRvcnkzNzE0MTk1OTg=",
      "_type": "github_account_owns_repo",
      "displayName": "OWNS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_key": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz|has|MDQ6VGVhbTQ4NTgxNjk=",
      "_toEntityKey": "MDQ6VGVhbTQ4NTgxNjk=",
      "_type": "github_account_has_team",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDQ6VGVhbTQ4NTgxNjk=",
      "_key": "MDQ6VGVhbTQ4NTgxNjk=|has|MDQ6VXNlcjUxMzUyMw==",
      "_toEntityKey": "MDQ6VXNlcjUxMzUyMw==",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDQ6VGVhbTQ4NTgxNjk=",
      "_key": "MDQ6VGVhbTQ4NTgxNjk=|has|MDQ6VXNlcjI1NDg5NDgy",
      "_toEntityKey": "MDQ6VXNlcjI1NDg5NDgy",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDQ6VGVhbTQ4NTgxNjk=",
      "_key": "MDQ6VGVhbTQ4NTgxNjk=|has|MDQ6VXNlcjYyNDkyMDk3",
      "_toEntityKey": "MDQ6VXNlcjYyNDkyMDk3",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjYyNDkyMDk3",
      "_key": "MDQ6VXNlcjYyNDkyMDk3|manages|MDQ6VGVhbTQ4NTgxNjk=",
      "_toEntityKey": "MDQ6VGVhbTQ4NTgxNjk=",
      "_type": "github_user_manages_team",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_key": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz|has|MDQ6VGVhbTQ4NTgxNzA=",
      "_toEntityKey": "MDQ6VGVhbTQ4NTgxNzA=",
      "_type": "github_account_has_team",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDQ6VGVhbTQ4NTgxNzA=",
      "_key": "MDQ6VGVhbTQ4NTgxNzA=|has|MDQ6VXNlcjYyNDkyMDk3",
      "_toEntityKey": "MDQ6VXNlcjYyNDkyMDk3",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjYyNDkyMDk3",
      "_key": "MDQ6VXNlcjYyNDkyMDk3|manages|MDQ6VGVhbTQ4NTgxNzA=",
      "_toEntityKey": "MDQ6VGVhbTQ4NTgxNzA=",
      "_type": "github_user_manages_team",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz",
      "_key": "MDEyOk9yZ2FuaXphdGlvbjg0OTIzNTAz|has|MDQ6VGVhbTQ4NTc0OTU=",
      "_toEntityKey": "MDQ6VGVhbTQ4NTc0OTU=",
      "_type": "github_account_has_team",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDQ6VGVhbTQ4NTc0OTU=",
      "_key": "MDQ6VGVhbTQ4NTc0OTU=|has|MDQ6VXNlcjI1NDg5NDgy",
      "_toEntityKey": "MDQ6VXNlcjI1NDg5NDgy",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "MDQ6VGVhbTQ4NTc0OTU=",
      "_key": "MDQ6VGVhbTQ4NTc0OTU=|has|MDQ6VXNlcjYyNDkyMDk3",
      "_toEntityKey": "MDQ6VXNlcjYyNDkyMDk3",
      "_type": "github_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "MDQ6VXNlcjYyNDkyMDk3",
      "_key": "MDQ6VXNlcjYyNDkyMDk3|manages|MDQ6VGVhbTQ4NTc0OTU=",
      "_toEntityKey": "MDQ6VGVhbTQ4NTc0OTU=",
      "_type": "github_user_manages_team",
      "displayName": "MANAGES",
    },
  ],
  "encounteredTypes": Array [
    "github_account",
    "github_user",
    "github_account_has_user",
    "github_user_manages_account",
    "github_repo",
    "github_account_owns_repo",
    "github_team",
    "github_account_has_team",
    "github_team_has_user",
    "github_user_manages_team",
  ],
  "numCollectedEntities": 8,
  "numCollectedRelationships": 17,
}
`;
